src/systems/a320_systems/src/flight_warning/mod.rs

```diff
@@ -0,0 +1,370 @@
+use systems::shared::{random_from_range, ElectricalBusType, ElectricalBuses};
+use systems::simulation::{
+    InitContext, Read, SimulationElement, SimulationElementVisitor, SimulatorReader,
+    SimulatorWriter, UpdateContext, VariableIdentifier, Write,
+};
+use uom::si::angle::degree;
+use uom::si::{f64::*, length::foot};
+
+mod ecam_control;
+mod input;
+mod runtime;
+
+/// This struct represents a physical flight warning computer, as installed on an A320.
+pub(super) struct A320FlightWarningComputer {
+    number: usize,
+    failure: Failure,
+
+    // Electrical Power
+    powered_by: ElectricalBusType,
+    is_powered: bool,
+
+    /// How long the FWC can tolerate a power loss and continue functioning afterwards without
+    /// interruption.
+    power_holdover: Duration,
+
+    /// How long the FWC has been unpowered for.
+    unpowered_for: Duration,
+
+    // How long the FWC takes to fully boot up and report useful outputs
+    self_test_time: Duration,
+
+    runtime: Option<A320FlightWarningComputerRuntime>,
+}
+
+impl A320FlightWarningComputer {
+    const MINIMUM_STARTUP_TIME_MILLIS: u64 = 50_000;
+    const MAXIMUM_STARTUP_TIME_MILLIS: u64 = 56_000;
+    const MINIMUM_POWER_HOLDOVER: u64 = 500;
+    const MAXIMUM_POWER_HOLDOVER: u64 = 1_500;
+
+    pub fn new(context: &mut InitContext, number: usize, powered_by: ElectricalBusType) -> Self {
+        let is_powered = context.has_engines_running();
+        Self {
+            number,
+            powered_by,
+            power_holdover: Duration::from_secs_f64(random_from_range(
+                Self::MINIMUM_POWER_HOLDOVER as f64 / 1000.,
+                Self::MAXIMUM_POWER_HOLDOVER as f64 / 1000.,
+            )),
+            self_test_time: Duration::from_secs_f64(random_from_range(
+                Self::MINIMUM_STARTUP_TIME_MILLIS as f64 / 1000.,
+                Self::MAXIMUM_STARTUP_TIME_MILLIS as f64 / 1000.,
+            )),
+            unpowered_for: if is_powered {
+                Duration::ZERO
+            } else {
+                Duration::from_millis(Self::MAXIMUM_POWER_HOLDOVER)
+            },
+            is_powered: false,
+            failure: Failure::new(FailureType::FlightWarningComputer(number)),
+            runtime: if is_powered {
+                Some(A320FlightWarningComputerRuntime::new_running())
+            } else {
+                None
+            },
+        }
+    }
+
+    fn update(&mut self, context: &UpdateContext, inputs: &A320FWCInputs) {
+        // Check for power and tick power capacitor holdover otherwise
+        if self.is_powered {
+            self.unpowered_for = Duration::ZERO;
+        } else {
+            self.unpowered_for += context.delta();
+        }
+
+        if self.unpowered_for < self.power_holdover && !self.failure.is_active() {
+            // Either initialize and run or continue running the existing runtime
+            let self_test_time = self.self_test_time;
```

Comment:
The assignment is needed because otherwise the borrow checker complains:
```
error[E0502]: cannot borrow `self.runtime` as mutable because it is also borrowed as immutable
  --> src\systems\a320_systems\src\flight_warning\mod.rs:91:27
   |
91 |               let runtime = self
   |  ___________________________^
92 | |                 .runtime
93 | |                 .get_or_insert_with(|| A320FlightWarningComputerRuntime::new(self.self_test_time));
   | |__________________------------------_--_______________________________________-------------------_^ mutable borrow occurs here
   |                    |                  |                                        |
   |                    |                  immutable borrow occurs here             first borrow occurs due to use of `self` in closure
   |                    immutable borrow later used by call
```
I think what happens is that:
- the borrow checker discovers that the closure tries to access self from the outer scope (albeit immutably)
- as such, the closure itself is marked as borrowing self (immutably)
- however, as the outer function (`get_or_insert_with`) needs a mutateble reference to `self.runtime`, we now have a conflicting mutable and immutable borrow (remember, there can be arbitrarily many immutable borrows, but only every single mutable borrow, and never both at the same time)

So there's two solutions:
- if we stick with get_or_insert_with, we can explicitly copy the duration and we can now safely use it inside the as there is no other reference to it. This happens with a simple assignment, as Duration implements `Copy`.
- we could also do something like `self.runtime = self.runtime.unwrap_or_else(<closure>);` to get rid of the simultaneous borrows, however as we need a `runtime` in the next line anyway to call `.update` we'd have to guard it again.

As such, I opted for the first solution with the seemingly redundant variable assignment. Unfortunately it hides the implicit Copy that I think is happening.